refactor(ui): project creation modal

MeBrei opened the pull request:
Closes #826 
Closes #835

rudolfs reviewed (commented):
Looks good!

Some things I noticed:

- [ ] circular deps
```
(!) Circular dependencies
ui/src/modal.ts -> ui/Screen/ProjectCreation.svelte -> ui/DesignSystem/Component/index.js -> ui/DesignSystem/Component/ModalOverlay.svelte -> ui/src/modal.ts
ui/src/modal.ts -> ui/Screen/ProjectCreation.svelte -> ui/DesignSystem/Component/index.js -> ui/DesignSystem/Component/Sidebar.svelte -> ui/src/modal.ts
```
- [ ] lock screen during create request broken
- [ ] keyboard shortcuts don't work when opening the modal via the "new project" button (might be related to https://github.com/radicle-dev/radicle-upstream/issues/826)

xla commented:
> ```
> (!) Circular dependencies
> ui/src/modal.ts -> ui/Screen/ProjectCreation.svelte -> ui/DesignSystem/Component/index.js -> ui/DesignSystem/Component/ModalOverlay.svelte -> ui/src/modal.ts
> ui/src/modal.ts -> ui/Screen/ProjectCreation.svelte -> ui/DesignSystem/Component/index.js -> ui/DesignSystem/Component/Sidebar.svelte -> ui/src/modal.ts
```

I ran into that as well when working on the navigation refactor. It might be good practice to separate the library code providing the machinery and the functionality mapping to actual components.

xla commented on ui/Screen/Profile/ProjectsMenu.svelte:
```diff
@@ -1,15 +1,16 @@
 <script>
-  import { push } from "svelte-spa-router";
-
   import * as path from "../../src/path.ts";
+  import * as modal from "../../src/modal.ts";
```
```suggestion
  import * as modal from "../../src/modal.ts";
  import * as path from "../../src/path.ts";
```

xla reviewed (changes requested):
I failed to bring this up during the original introduction of the new modal, but I don't see a reason why we have to hold onto the `path` concept even so we don't use a router anymore. An enum for our differrent modals could look like this:

``` ts
// app.ts

import { SvelteComponent } from "svelte";

import Help from "../Modal/Help.svelte";
import ProjectCreation from "../Modal/ProjectCreation.svelte";
import Shortcuts from "../Modal/Shortcuts.svelte";

export interface View<K extends string> {
  readonly component: unknown;
  readonly key: K,
  readonly props?: Props;
}

export type Props = Record<string, 0 | string>;
export type Map<Key extends string, C extends typeof SvelteComponent> = Required<
  Record<Key, C>
>;

export enum Modal {
  Help = "Help",
  ProjectCreation = "ProjectCreation",
  Shortcuts = "Shortcuts",
}

export const modalMap: Map<Modal, typeof SvelteComponent> = {
  [Modal.Help]: Help,
  [Modal.ProjectCreation]: ProjectCreation,
  [Modal.Shortcuts]: Shortcuts,
};
```

MeBrei commented:
Closes #826 

3ec3d9a should also fix the issue above

MeBrei commented:
> I failed to bring this up during the original introduction of the new modal, but I don't see a reason why we have to hold onto the `path` concept even so we don't use a router anymore. An enum for our differrent modals could look like this:
> ...

I refactored the setup of the modals in 511155f36c5cbfd56b57632611fa8b9e11b1bd3a
I kept the routes for now, this way, I like that everything is set up the same (ie routes map in `App.svelte` and the routes being accessible via `path.ts`). Having the mapping of the routes in `App.svelte` also avoids the circular dependencies. I also moved the modals into `Screen` as some of them have advanced functionality (eg API calls) that don't fit into `Component` and the modals basically form a screen in themselves.

xla commented:
> I kept the routes for now, this way, I like that everything is set up the same (ie routes map in `App.svelte` and the routes being accessible via `path.ts`). Having the mapping of the routes in `App.svelte` also avoids the circular dependencies. I also moved the modals into `Screen` as some of them have advanced functionality (eg API calls) that don't fit into `Component` and the modals basically form a screen in themselves.

The consistency argument is a good one and I think we can drop the route concept once we migrate away from router based navigation altogether in #754.

I also believe that modals should be a first class concept and be treated like screens in the hierarchy and in their allowed capabilities/access. It feels strange to move them into the screen directory, I'd put them `ui/Modal`. So we can have:

```
./
../
ui/
ui/Modal/MyModal.svelte
ui/Modal/MyModal/Fragment1.svelte
ui/Modal/MyModal/Fragment2.svelte
ui/Screen/MyScreen.svelte
ui/Screen/MyScreen/Fragment1.svelte
ui/Screen/MyScreen/Fragment2.svelte
```

xla reviewed (approved):
🌙 🎁 🕧 🎸

rudolfs commented:
Found some edge cases, that don't make sense after the reshuffling of the inputs in the form. Feel free to create a follow up issue if it's too much hassle to address in this PR.

- [ ] when choosing "create new repo", after selecting the directory path it should focus on the name field instead of immediately showing a validation error
<img width="562" alt="Screenshot 2020-09-04 at 10 50 28" src="https://user-images.githubusercontent.com/158411/92221180-5e719680-ee9d-11ea-89ec-76a3649ee600.png">

- [ ] when there's only one character in the name field, and the validations don't pass, the "create project" button is already enabled, it should stay disabled until all validations pass
<img width="561" alt="Screenshot 2020-09-04 at 10 50 11" src="https://user-images.githubusercontent.com/158411/92221223-721cfd00-ee9d-11ea-8379-fafc50b03319.png">

MeBrei commented:
> Found some edge cases, that don't make sense after the reshuffling of the inputs in the form. Feel free to create a follow up issue if it's too much hassle to address in this PR.
> 

I think it would make sense to refactor the validation set up for this modal before addressing the edge cases, so I created a new issue #869
